sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -85,10 +85,42 @@ object OrcUtils extends Logging {
   }
 
   private def toCatalystSchema(schema: TypeDescription): StructType = {
+    import TypeDescription.Category
+
+    def toCatalystType(orcType: TypeDescription): DataType = {
+      orcType.getCategory match {
+        case Category.STRUCT => toStructType(orcType)
+        case Category.LIST => toArrayType(orcType)
+        case Category.MAP => toMapType(orcType)
+        case _ => CatalystSqlParser.parseDataType(orcType.toString)
+      }
+    }
+
+    def toStructType(orcType: TypeDescription): StructType = {
+      val fieldNames = orcType.getFieldNames.asScala
+      val fieldTypes = orcType.getChildren.asScala
+      fieldNames.zip(fieldTypes).foldLeft(new StructType) {
+        case (resultType, (fieldName, fieldType)) =>
+          val catalystType = toCatalystType(fieldType)
+          resultType.add(StructField(fieldName, catalystType))
```

Comment:
does orc struct have nullability and metadata?

Reply:
All the types in ORC seems to have nullability.
https://orc.apache.org/docs/types.html

Schema mapping from Catalyst to ORC seems to be done using `OrcFileFormat.getQuotedSchemaString`, and it doesn't consider only `name` and `dataType` so I think we don't need to care about metadata here.

https://github.com/apache/spark/blob/ff7705ad2ad5f4b9dfbeda83e93a0db676e1ffd9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala#L47-L59